Overloaded augmented assignments

japaric opened the pull request:
Implements overload-able augmented/compound assignments, like `a += b` via the `AddAssign` trait, as specified in RFC [953]

[953]: https://github.com/rust-lang/rfcs/blob/master/text/0953-op-assign.md

r? @nikomatsakis

bors commented:
:umbrella: The latest upstream changes (presumably #28358) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented on src/librustc_typeck/check/op.rs:
```diff
@@ -167,14 +155,16 @@ fn check_overloaded_binop<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
                                     rhs_expr: &'tcx hir::Expr,
-                                    op: hir::BinOp)
+                                    op: hir::BinOp,
+                                    assign: bool)
```
Nit: can we newtype this bool? I think I prefer something like:

```
struct IsAssign(bool)
```

and then in the call `foo(..., IsAssign(true))`, though `IsAssign::Yes` is also ok.

nikomatsakis commented on src/librustc_typeck/check/op.rs:
```diff
@@ -231,27 +228,51 @@ pub fn check_user_unop<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
-fn name_and_trait_def_id(fcx: &FnCtxt, op: hir::BinOp) -> (&'static str, Option<DefId>) {
+fn name_and_trait_def_id(fcx: &FnCtxt,
+                         op: hir::BinOp,
+                         assign: bool)
```
Nit: newtype here too.

nikomatsakis commented on src/librustc_typeck/check/writeback.rs:
```diff
@@ -90,24 +90,51 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
-                    self.fcx.inh.tables.borrow_mut().adjustments.remove(&lhs.id);
+        match e.node {
+            hir::ExprBinary(ref op, ref lhs, ref rhs) |
+                hir::ExprAssignOp(ref op, ref lhs, ref rhs) => {
```
nit: indentation looks off

nikomatsakis commented on src/librustc_typeck/check/writeback.rs:
```diff
@@ -90,24 +90,51 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
+
+                        if let hir::ExprAssignOp(..) = e.node {
+                            if !tcx.sess.features.borrow().augmented_assignments &&
+                                !self.fcx.expr_ty(e).references_error() {
```
Nit: the body of the if and the condition ought not be aligned the same without an intervening `{`, imo. :) Can you move the opening brace to the next line, or else modify the indent, or something? I personally often do this but I think I am the only one that does so: ;)

```rust
if
    cond1 &&
    cond2
{
    ...
}
```

nikomatsakis commented:
This looks good. r+ modulo nits -- can you also add a test that you can't implement `AddAssign` and friends without the suitable feature gate? (Another related question -- do we really want two distinct feature gates? Seems like the traits and associated syntax live or die together?)

Also: I think it may be worth waiting to land this until the beta is released, just in case it introduces accidental breakage.

japaric commented:
@nikomatsakis I've addressed all your comments (I think)

> do we really want two distinct feature gates? Seems like the traits and associated syntax live or die together?

I did try using the same feature name for the language feature and for the traits, but it didn't work. What I observed is that when I added `#![feature(augmented_assignment)]` to some source file, the language feature was enabled, i.e. `a += b` didn't raised an error; but implementing the `OpAssign` trait still raised the "unstable library feature 'augmented_assignments'" error. Which seems to indicate there's a bug when a language feature and a library feature have the same name: `#![feature(foo)]` enables the language feature `foo`, but not the library feature `foo` when it should enable both.

> Also: I think it may be worth waiting to land this until the beta is released, just in case it introduces accidental breakage.

Sounds good to me

bors commented:
:umbrella: The latest upstream changes (presumably #28336) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
@japaric ok feel free to r=me once rebased

japaric commented:
@bors r=nmatsakis

bors commented:
:pushpin: Commit f5569ec has been approved by `nmatsakis`

<!-- @bors r=nmatsakis f5569ecd7682a22f9ae3293a89479c7b99b6d941 -->

bors commented:
:hourglass: Testing commit f5569ec with merge c61bf60...

bors commented:
:broken_heart: Test failed - [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/6464)

bluss commented:
@bors retry

bors commented:
:hourglass: Testing commit f5569ec with merge 295d51f...

bors commented:
:broken_heart: Test failed - [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/6468)

alexcrichton commented:
@bors: retry

On Sat, Sep 19, 2015 at 4:08 AM, bors <notifications@github.com> wrote:

> [image: :broken_heart:] Test failed - auto-linux-64-opt
> <http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/6468>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/28345#issuecomment-141650702>.
>

bors commented:
:hourglass: Testing commit f5569ec with merge 783c3fc...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/6496), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/6510), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/6449), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/6475), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/6453), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/6529), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/6523), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/6487), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/1472), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/1471), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/1471), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/1473), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/750), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/1454)
